### PR TITLE
remote-build: autorecovery

### PR DESCRIPTION
**Pull request by cjp256**

### CODE REVIEW

**Comment by cjp256** [comment]
Admittedly https://github.com/snapcore/snapcraft/pull/2768/commits/4cc50a8d6931b8cac3394e9894c5396ac0c3f25b looks a little worse than it is because I foolishly shuffled some methods around... I can break up that commit if it helps :)

File: snapcraft/internal/remote_build/_launchpad.py [hunk 1]
```diff
@@ -112,13 +112,18 @@ def _create_cache_directory(self) -> str:
...
-        )
+        try:
+            self._lp = Launchpad.login_with(
+                "snapcraft remote-build {}".format(snapcraft.__version__),
+                "production",
+                self._cache_dir,
+                credentials_file=self._credentials,
+                version="devel",
+            )
+        except ConnectionRefusedError:
```
Comment (sergiusens) [inline snapcraft/internal/remote_build/_launchpad.py]: `except (ConnectionRefusedError, TimeoutError):`

File: snapcraft/cli/remote.py [hunk 2]
```diff
@@ -235,3 +208,27 @@ def _get_project_architectures(project) -> List[str]:
...
+    project_architectures = _get_project_architectures(project)
+    if project_architectures and user_specified_arch:
+        raise click.BadOptionUsage(
+            "Cannot use --arch, architecture list is already set in snapcraft.yaml."
+        )
+
+    if project_architectures:
+        archs = project_architectures
+    elif user_specified_arch == "all":
+        archs = _SUPPORTED_ARCHS
```
Comment (sergiusens) [inline snapcraft/cli/remote.py]: your refactor made me notice this, please remove this option `all` has a very different meaning in the `architectures` list in `snapcraft.yaml`.

File: snapcraft/internal/remote_build/_launchpad.py [hunk 3]
```diff
@@ -179,176 +210,115 @@ def create_snap(self, force=False) -> Entry:
...
+    def _create_git_repository(self, force=False) -> Entry:
+        """Create git repository."""
+        if force:
+            self._delete_git_repository()
+
+        logger.debug("Creating source repository on Launchpad...")
+        logger.debug(
+            f"name={self._lp_name} owner={self._lp_owner} target={self._lp_owner}"
+        )
+        return self._lp_post_new_git_repository()
```
Comment (sergiusens) [inline snapcraft/internal/remote_build/_launchpad.py]: this name doesn't go easy create ends up posting a new repo?

File: snapcraft/cli/remote.py [hunk 4]
```diff
@@ -150,6 +156,12 @@ def remote_build(
     _monitor_build(lp)
 
 
+def _clean_build(lp: LaunchpadClient):
+    echo.info("Cleaning existing builds and artifacts...")
```
Comment (sergiusens) [inline snapcraft/cli/remote.py]: s/builds/tasks/ or jobs

File: snapcraft/cli/remote.py [hunk 5]
```diff
@@ -99,6 +102,11 @@ def remote_build(
         "snapcraft remote-build is experimental and is subject to change - use with caution."
     )
 
+    if debug:
+        logging.getLogger("snapcraft.internal.remote_build._launchpad").setLevel(
```
Comment (sergiusens) [inline snapcraft/cli/remote.py]: Please live this to the developer env we have or if this is user facing, then let's punt this until we have a better design for general logging wrt verbosity

File: snapcraft/internal/remote_build/_launchpad.py [hunk 6]
```diff
@@ -36,7 +36,7 @@
 from snapcraft.internal.sources._git import Git
 from snapcraft.project import Project
 
-_LP_POLL_INTERVAL = 30
+_LP_POLL_INTERVAL = 5
```
Comment (sergiusens) [inline snapcraft/internal/remote_build/_launchpad.py]: Is Colin ok with this?

File: snapcraft/internal/remote_build/_launchpad.py [hunk 7]
```diff
@@ -468,6 +468,12 @@ def push_source_tree(self, repo_dir: str) -> str:
 
         logger.info(f"Sending build data to Launchpad... ({stripped_url})")
 
+        # Undocumented environment variable to show the URL with secrets,
+        # if it is really needed to reproduce an issue (as there is no
+        # other way for user to recreate an access token manually.)
+        if os.getenv("SNAPCRAFT_DEBUG_SECRETS"):
+            logger.debug(f"({url})")
```
Comment (sergiusens) [inline snapcraft/internal/remote_build/_launchpad.py]: Let's not do this

File: snapcraft/internal/remote_build/_launchpad.py (hunk 1 above)
Comment (cjp256) [reply snapcraft/internal/remote_build/_launchpad.py]: Will do!

File: snapcraft/cli/remote.py (hunk 2 above)
Comment (cjp256) [reply snapcraft/cli/remote.py]: Remove the `all` option altogether?

**Comment by cjp256** [comment]
Pulled https://github.com/snapcore/snapcraft/pull/2772 out of this PR, fixing the above comments and adding tests.

File: snapcraft/internal/remote_build/_launchpad.py (hunk 1 above)
Comment (cjp256) [reply snapcraft/internal/remote_build/_launchpad.py]: Done: https://github.com/snapcore/snapcraft/pull/2772

File: snapcraft/cli/remote.py (hunk 2 above)
Comment (cjp256) [reply snapcraft/cli/remote.py]: Pulled out arch handling into its own PR (and fixed): https://github.com/snapcore/snapcraft/pull/2774/commits

File: snapcraft/cli/remote.py (hunk 4 above)
Comment (cjp256) [reply snapcraft/cli/remote.py]: Pulled out and simplified to "Cleaning..." https://github.com/snapcore/snapcraft/pull/2775

File: snapcraft/internal/remote_build/_launchpad.py (hunk 3 above)
Comment (cjp256) [reply snapcraft/internal/remote_build/_launchpad.py]: Fixed.

File: snapcraft/internal/remote_build/_launchpad.py (hunk 6 above)
Comment (cjp256) [reply snapcraft/internal/remote_build/_launchpad.py]: Removed.

File: snapcraft/internal/remote_build/_launchpad.py (hunk 7 above)
Comment (cjp256) [reply snapcraft/internal/remote_build/_launchpad.py]: Removed.

**Comment by cjp256** [comment]
Requires https://github.com/snapcore/snapcraft/pull/2772

File: snapcraft/cli/remote.py (hunk 5 above)
Comment (cjp256) [reply snapcraft/cli/remote.py]: Removed

File: snapcraft/cli/remote.py [hunk 8]
```diff
@@ -136,14 +122,30 @@ def remote_build(
...
+
+        echo.info("Recovering build...")
+        _monitor_build(lp)
+    elif lp.has_outstanding_build():
+        # There was a previous build that hasn't finished.
+        # Recover from interrupted build.
+        echo.info("Found previously started build.")
+        _print_status(lp)
+
+        if not click.confirm("Do you wish to recover this build?", default=True):
```
Comment (sergiusens) [inline snapcraft/cli/remote.py]: should use echo.confirm

File: snapcraft/internal/remote_build/_launchpad.py [hunk 9]
```diff
@@ -14,20 +14,20 @@
...
 import gzip
 import logging
 import os
 import shutil
 import time
 import urllib.request
 import urllib.error
-import urllib.parse
 
+from datetime import datetime
```
Comment (sergiusens) [inline snapcraft/internal/remote_build/_launchpad.py]: we should discuss imports. We also follow PEP8 for this https://www.python.org/dev/peps/pep-0008/#imports

File: snapcraft/cli/remote.py (hunk 8 above)
Comment (cjp256) [reply snapcraft/cli/remote.py]: Will do

Comment (cjp256) [reply snapcraft/cli/remote.py]: Done!

File: snapcraft/internal/remote_build/_launchpad.py (hunk 9 above)
Comment (cjp256) [reply snapcraft/internal/remote_build/_launchpad.py]: Fixed!

**Comment by cjp256** [comment]
addressed comments (first force push) then rebased (second force push)

**Pull request merged**